### PR TITLE
[megatron, doc] feat: add SAPO Megatron NPU training script and deployment guide for Qwen3.5-4B

### PR DESCRIPTION
**Pull request by 5082459**
### What does this PR do?

Adds the first **SAPO + Megatron** training script for **Ascend NPU**, targeting the **Qwen3.5-4B** dense model with Gated Delta Net (GDN) linear attention. The existing SAPO scripts (`examples/sapo_trainer/`) only cover FSDP2 on GPU; this PR fills the gap for Megatron backend on NPU.

Two files are added:

- `examples/ascend_extras/sapo_trainer/run_qwen3_5_4b_megatron_npu.sh` — Megatron backend + vLLM rollout, tested on 8 nodes × 16 NPUs (128 total) with TP=4 PP=2 CP=1 GEN_TP=4 (DP=16).
- `docs/ascend_tutorial/zh/model_support/examples/sapo_qwen3_5_4b_megatron_npu.md` — Full deployment practice with environment setup, Megatron-Bridge installation guide, mermaid flowchart + sequence diagram, and 4 troubleshooting Q&As.

### Checklist Before Starting

- [x] Search for similar PRs: https://github.com/verl-project/verl/pulls?q=is%3Apr+sapo+megatron+npu — no existing SAPO + Megatron + NPU script found. Existing SAPO scripts are FSDP-only (`examples/sapo_trainer/run_qwen3_8b_fsdp.sh`, `run_qwen3_30b_a3b_fsdp.sh`). Existing NPU Megatron scripts use GRPO (`examples/ascend_extras/grpo_trainer/`) or GSPO (`examples/ascend_extras/gspo_trainer/`), not SAPO.
- [x] Format the PR title as `[{modules}] {type}: {description}` — `[megatron, doc] feat: ...`

### Test

**Not CI-testable** — This script requires Ascend NPU hardware (8 nodes × 16 NPUs = 128 NPUs) and is validated by real deployment, not unit tests.

Tested configuration:

| Item | Value |
|---|---|
| Model | Qwen3.5-4B (dense, GDN linear attention) |
| Algorithm | SAPO (`loss_mode=sapo`, `tau_pos=1.0`, `tau_neg=1.05`) |
| Backend | Megatron (via Megatron-Bridge `de93536e`) + vLLM rollout |
| Hardware | 8 nodes × 16 NPUs = 128 NPUs (Atlas 800T A3) |
| Parallelism | TP=4 PP=2 CP=1 EP=1 GEN_TP=4 → DP=16 |
| Dataset | DAPO-Math-17k |
| train_batch_size | 512, rollout.n=16 |

### API and Usage Example

```bash
# 1. Start Ray cluster (head node)
ray start --head --port 6766 --resources='{"NPU": 16}'
ray status

# 2. Worker nodes join (remaining 7 nodes)
ray start --address=<head_ip>:6766 --resources='{"NPU": 16}'

# 3. Launch SAPO training
bash examples/ascend_extras/sapo_trainer/run_qwen3_5_4b_megatron_npu.sh

# Override defaults via env vars
MODEL_PATH=/path/to/Qwen3.5-4B \
TRAIN_FILE=/path/to/train.parquet \
VAL_FILE=/path/to/val.parquet \
TP=4 PP=2 \
bash examples/ascend_extras/sapo_trainer/run_qwen3_5_4b_megatron_npu.sh
```

### Design & Code Changes

**Script** (`run_qwen3_5_4b_megatron_npu.sh`):

- Entry point: `verl.trainer.main_ppo` with `model_engine=megatron`
- SAPO loss: `loss_mode=sapo` with `tau_pos=1.0` / `tau_neg=1.05` (asymmetric gating)
- GDN architecture constraint: `use_remove_padding=False` + `use_dynamic_bsz=False` (forces bshd compute format — Qwen3.5 GDN does not support packed sequences / THD format in Megatron-LM)
- Megatron-Bridge integration: `use_mbridge=True, vanilla_mbridge=False` (official NVIDIA-NeMo/Megatron-Bridge, not the deprecated ISEEKYAN/mbridge)
- Full offload: param/optimizer/grad offload + precision-aware optimizer + CPU overlap
- Style aligned with existing `examples/ascend_extras/grpo_trainer/run_qwen3_5_122b_a10b_32k_megatron.sh`

**Documentation** (`sapo_qwen3_5_4b_megatron_npu.md`):

- Algorithm config with SAPO parameter explanation
- Qwen3.5 GDN architecture constraint section (why `use_remove_padding=False` is required)
- Megatron-Bridge installation guide (disambiguating `mbridge` vs `Megatron-Bridge` — two different packages with the same import name `megatron_bridge`)
- Key training parameters table (batch sizes, lengths, offload, transformer config overrides)
- Mermaid flowchart (training loop) + sequence diagram (per-step data flow)
- 4 troubleshooting Q&As from real deployment:
  - **Q1**: Missing Megatron-Bridge (Docker image ships deprecated `mbridge` that doesn't support Qwen3.5)
  - **Q2**: `mstx.range_end` signature mismatch (MindSpeed NVTX→MSTX patch incompatibility)
  - **Q3**: Checkpoint global shape mismatch (stale checkpoint from different model/parallelism config)
  - **Q4**: Transformers 5.x tokenizer incompatibility (`TokenizersBackend` class not found in 4.x)

### Checklist Before Submitting

- [x] Read the [Contribute Guide](https://github.com/verl-project/verl/blob/main/CONTRIBUTING.md).
- [ ] Apply pre-commit checks: `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always` — pre-commit not fully run locally (hook installation timed out on this machine). Manual checks (trailing whitespace, EOF newline) passed. Will fix any CI failures reported by the bot.
- [x] Add / Update the documentation — added `docs/ascend_tutorial/zh/model_support/examples/sapo_qwen3_5_4b_megatron_npu.md`.
- [ ] Add unit or end-to-end test(s) to the CI workflow — not feasible: SAPO + Megatron + NPU requires physical Ascend hardware (128 NPUs). No existing CI workflow covers this combination. The script itself serves as the reproducible end-to-end test.
- [ ] Once PR is ready for CI, send a message in the `ci-request` channel in the verl Slack workspace.
- [ ] This PR does not touch the `recipe` submodule.

### CODE REVIEW

**Comment by wucong25** [comment]
建议合入到verl-ascend-recipe仓库中
